sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/GenericArrayData.scala

```diff
@@ -142,3 +164,415 @@ class GenericArrayData(val array: Array[Any]) extends ArrayData {
     result
   }
 }
+
+final class GenericIntArrayData(private val primitiveArray: Array[Int])
```

Comment:
I think the class hierarchy is a bit off. We should make `GenericArrayData` abstract and introduce a `GenericeRefArrayData` as a replacement for the current `GenericArrayData`.

The downside of the introducing this many `ArrayData` classes is that `ArrayData` operations will be mega-morphic.